chapters/io/index.rst

```diff
@@ -0,0 +1,315 @@
+
+    .. highlight:: python
+    .. code-block:: python
+
+        with f:
+            f.index_csv(open('input2.csv'), field_resolver={'identifier': 'id', 'paragraph': 'text'})
+
+Output
+------
+
+The output of flow operations is handled via callback functions `on_done`, `on_error` and `on_always`.
+
+    .. highlight:: python
+    .. code-block:: python
+
+        def handle_response(resp):
+            # the response handler is only called if the flow execution does not encounter exceptions
+            # handle the response
+            for d in resp.search.docs:
+                ...
+                for m in d.matches:
+                   ...
+
+
+        def handle_error():
+            # in case of an Exception, the flow execution continues and calls this `on_error` handler
+
+        def handle_search_done():
+            # this handler is always called regardless of Exceptions
+
+        with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+            f.search([doc], on_done=handle_response, on_error=handle_error, on_always=handle_search_done)
+
+
+It can be handy to use the built in `print` function as `on_done` callback.
+    .. highlight:: python
+    .. code-block:: python
+
+        with f:
+            f.search(input_fn, on_done=print)
+
+    
+Insights
+--------
+
+When using the `Flow.*` functions, Jina builds and sends Protobuf messages to the relevant pods.
+For instance calling the `index_ndarray(...)` function sends the following message to the first Pod.
+
+```
+request {
+  request_id: 1
+  index {
+    docs {
+      id: 1
+      weight: 1.0
+      length: 100
+      blob {
+        buffer: "\004@\316\362/D\333?\244>\235\305\027\311\336?\267\210\251\311^\260\345?\366\n(\014\022m\356?\374\262\017\030\036\357\351?-c\300\337\217V\345?\241G\241\352\233\024\356?\340\346lUf\353\350?"
+        shape: 8
+        dtype: "float64"
+      }
+    }
+    docs {
+      id: 2
+      weight: 1.0
+      length: 100
+      blob {
+        buffer: "\312Wm\337\250\217\354?t\212\326\020\261\r\320?\254\262\300u<O\323?\340\210\222$\321\216\314?\310.q,+\347\311?&\316\361\310\252R\331?\214\016\201a\231\262\330?\342\231\262\221\343%\324?"
+        shape: 8
+        dtype: "float64"
+      }
+    }
+    docs {
+      id: 3
+      weight: 1.0
+      length: 100
+      blob {
+        buffer: "kT\250\372K%\345?\237\017+u\300\227\353?\3668\256\340\251\227\350?\327\006$\032$\002\341?\274\300\3573\371\262\343?\346\371\265dV\330\342?\370\210\360\002P3\340?\022i-\016\374\320\331?"
+        shape: 8
+        dtype: "float64"
+      }
+    }
+  }
+}
+```
+
+The structure of this message is defined in the format of [protobuf](https://docs.jina.ai/chapters/proto/docs.html).
+Check more details of the data structure at [`jina.proto`](https://docs.jina.ai/chapters/proto/docs.html#jina.proto).
+
+`request` contains input data and related metadata.
+The input is a 3*8 matrix that is sent to the Flow, which matches 3 `request.index.docs`,
```

Comment:
I think the `*` is confusing the renderer. Might be the same for rst. Check how it looks

Reply:
good notion. Seems like we are lucky. It is rendered like a charm.